### PR TITLE
[SPARK-20887][CORE] support alternative keys in ConfigBuilder

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

`ConfigBuilder` builds `ConfigEntry` which can only read value with one key, if we wanna change the config name but still keep the old one, it's hard to do.

This PR introduce `ConfigBuilder.withAlternative`, to support reading config value with alternative keys. And also rename `spark.scheduler.listenerbus.eventqueue.size` to `spark.scheduler.listenerbus.eventqueue.capacity` with this feature, according to https://github.com/apache/spark/pull/14269#discussion_r118432313

## How was this patch tested?

a new test

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @JoshRosen @dhruve

**Comment by JoshRosen** [comment]
FWIW I didn't actually say that we should rename that key since the cost of
the confusing name isn't that high right now. So while I don't oppose this
mechanism I'm neutral on it given that the only use case so far seems kind
of minor. I was mostly commenting just so that future readers and reviewers
can more easily spot the issue and hopefully pick better names going
forward.

**Comment by cloud-fan** [comment]
@JoshRosen actually there are more use cases. Currently we have 2 ways to define a config: create a config entry in the `config` package object, or `SQLConf` object, and read conf value with the config entry in code. Or read conf value with hard-coded conf name directly in the code. Ideally we should always create config entry, but we can't because it doesn't support alternative keys.

**Comment by JoshRosen** [comment]
@cloud-fan, what about `SparkConf`'s `configsWithAlternatives `: https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/SparkConf.scala#L596

**Comment by cloud-fan** [comment]
It's only used in the `SparkCong.get(key: String)` code path, not `SparkConf.get(entry: ConfigEntry[T])` code path. That's why we only support alternative keys if users get conf value by hard-coded conf name.

**Comment by JoshRosen** [comment]
Ahhh, makes sense. Thanks for the clarification.

File: core/src/main/scala/org/apache/spark/internal/config/ConfigReader.scala [hunk 1]
```diff
@@ -102,4 +102,16 @@ private[spark] class ConfigReader(conf: ConfigProvider) {
     }
   }
 
+  private def getOrDefault(conf: ConfigProvider, key: String): Option[String] = {
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/internal/config/ConfigReader.scala]: nit: add comment for this method.

**Review by jiangxb1987 — APPROVED** [review]
LGTM only have a nit.

**Comment by jiangxb1987** [comment]
restest please

**Comment by jiangxb1987** [comment]
retest please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks for the review, merging to master!

**Pull request closed without merging**